doc/man/man1/lsc.1
.TH lsc 1 User Manuals
.SH NAME
lsc \- LSC command
.SH SYNOPSIS
\fB $LSC_HOME/bin/lsc or $LSC_HOME/bin/lsc.cmd \f1
.SH DESCRIPTION
lsc is the command line that will launch LSC synchronization engine. It may be a run-and-stop or a run-and-block command depending on the synchronization mode. Connections and synchronization options are defined in \fBlsc.xml(5)\f1 file. 

Ldap Synchronization Connector, or LSC, is a tool for synchronizing data repositories, with a particular focus on identity information. 

It can read from a variety of different sources, perform limitless transformations of data on-the-fly and write the results to a LDAP directory. The tool is designed to be fast, and extensible in all areas. 

It is an open source project, released under the BSD license.

Requirements: JRE 1.6. Most tests are run against Sun / Oracle JRE but basic build is also achieved through OpenJDK 6. 
.SH GENERAL DESCRIPTION
The LSC XML file format is directly map with configuration objects. The settings full description is included as the Javadoc class comments. You will find there a shorter version excluding plugins. 

Each parameter can be either a static or a dynamic value. If dynamic the value is interpreted as a scripting string, Javascript by default or any JSR 223 supported language. 

In the XML file, you will find 3 main sections : - tasks - connections - security settings 

The tasks described how the source service is synchronized with the destination service. It mainly contains specific services settings (LDAP filter, SQL request, ...) and synchronization options (mapping from a dataset to another). The connections zone contains the description of the basic services settings (url, login, password, ...). The security zone includes severeals shared settings about encryption, hashing, ... 

This documentation is organized in a "A.B.C = TEXT" format that must be understood in XML notation like : <lsc> <A> <B> <C>TEXT</C> </B> </A> </lsc> The lsc xml node is not included in the following options description but is implied and mandatory </p> <p> The id attribute that may be included inside node is forced by the XML to object mapping engine (XStream) and does not have any particular meaning. If you manipulate the XML file format by your own, just remove it, XStream will do the job. Take core not to copy it to avoid giving the same id to two different nodes. 
.SH TASK SECTION
.TP
\fBtask.name\f1
Each task must be uniquely named to identify it inside the engine and it inside log messages. MANDATORY - STATIC 
.TP
\fBtask.bean\f1
This option describes the Java object i OPTIONAL - STATIC - Default to "org.lsc.beans.SimpleBean" 
.TP
\fBdn.real_root\f1
LSC also has to know the real root DN (used to build factice DN for service as database for example). You should specify a complete LDAP DN, like \fBdn.real_root = dc=lsc-project,dc=org\f1 for example. 
.TP
\fBdn.people\f1
LSC also has to know the RDN of the people's tree. You should specify a LDAP RDN, like \fBdn.people = ou=people\f1 for example. 
.TP
\fBlsc.tasks\f1
Name of available LSC tasks, separated by commas. For example, you could have \fBtask1, task2, task3\f1 as a value. Be careful that tasks also have to be defined. 
.SH SERVICE CONFIGURATION OPTIONS
These are global configuration options to define the source and destination services. You must replace the term \fBX\f1 in option label by \fBsrc\f1 to configure source service, or \fBdst\f1 to configure destination service. 
.TP
X\fB.java.naming.factory.initial\f1
Initial context factory to use for data abstraction layer. You could specify the value \fBcom.sun.jndi.ldap.LdapCtxFactory\f1. It should be the same value for both LDAP source and destination services. 
.TP
X\fB.java.naming.ldap.version\f1
Version of the LDAP protocol to use to connect to the LDAP server. \fB3\f1 is recommended. 
.TP
X\fB.java.naming.provider.url\f1
URI LDAP to connect to. Note that specifying the search base in the URI is not necessary, but could improve performance. 
.TP
X\fB.java.naming.security.authentication\f1
LDAP authentication mechanism to use. Generally, this value does not need to be changed. Put \fBsimple\f1 value to use authentication based on DN and password. 
.TP
X\fB.java.naming.security.principal\f1
LSC uses a DN on the LDAP server to authenticate itself and get rights on LDAP data. The value is a complete DN of an existing entry in the specified LDAP server. 
.TP
X\fB.java.naming.security.credentials\f1
The appropriate password for the DN used above, in case you use simple LDAP authentication. 
.TP
X\fB.java.naming.referral\f1
Choose how to follow LDAP referrals. Values could be \fBignore\f1, \fBfollow\f1, or \fBthrow\f1. 
.TP
X\fB.java.naming.ldap.derefAliases\f1
Choose how to follow LDAP aliases. Values could be \fBnever\f1, \fBalways\f1, \fBsearch\f1 or \fBfind\f1. 
.SH TASK CONFIGURATION OPTIONS
This section describes available options to configure a LSC task. You should replace \fBTASKNAME\f1 by a task name specified in the global configuration option \fBlsc.tasks\f1. 
.TP
\fBlsc.tasks.\f1TASKNAME\fB.object\f1
The full name of the JAVA class to use for plain LDAP object. For example, for person, you could use \fBorg.lsc.objects.inetOrgPerson\f1. Be careful that this class has to exist in this instance of LSC. 
.TP
\fBlsc.tasks.\f1TASKNAME\fB.bean\f1
The full name of the JAVA BEAN class to use for specialisation of the synchronization. For example, for person, you could use \fBorg.lsc.beans.inetOrgPersonBean\f1. Be careful that this class has to exist in this instance of LSC. 
.TP
\fBlsc.tasks.\f1TASKNAME\fB.type\f1
The synchronisation type for this task between the two defined services. 
.TP
\fBlsc.tasks.\f1TASKNAME\fB.srcService\f1
The fullname of the JAVA JNDI class to use to retrieve information from source service. There is a generic class for LDAP connection, which is \fBorg.lsc.jndi.SimpleJndiSrcService\f1. 
.TP
\fBlsc.tasks.\f1TASKNAME\fB.srcService.baseDn\f1
The RDN of the people tree on the source service. Generally, it should be \fBou=people\f1. 
.TP
\fBlsc.tasks.\f1TASKNAME\fB.srcService.attrId\f1
The attribute used to identify a user in the source service. This attribute must be used in the identity LDAP filter (see filterId below). For example, if users have a unique value of the LDAP attribute "uid", then you could use the value \fBuid\f1 here. 
.TP
\fBlsc.tasks.\f1TASKNAME\fB.srcService.filterAll\f1
This is the "global LDAP filter" used to retrieve all user DNs from the source service. 
.TP
\fBlsc.tasks.\f1TASKNAME\fB.srcService.filterId\f1
This is the "identity LDAP filter" used to retrieve one entry from the source service. You have to use the string \fB{0}\f1 in the filter. This string represents a user identifier found in user entry returned by the global LDAP filter. So, for example, a filter could be \fB(&(objectclass=inetOrgPerson)(mail=*)(uid={0}))\f1
.TP
\fBlsc.tasks.\f1TASKNAME\fB.srcService.attrs\f1
Here, you specify all returned attributes from the source service for one LDAP entry. Generally, these attributes will be used to build the new SASL userPassword. Be careful that the value of this option has to be same as in the source code of this instance of LSC. In fact, the algorithm used is in the JAVA BEAN, in the method named by attributes specified here. So, it strictly not recommended to modify this value (\fBloginName userPassword\f1). 
.TP
\fBlsc.tasks.\f1TASKNAME\fB.dstService\f1
The fullname of the JAVA JNDI class to use to retrieve information from destination service. There is a generic class for LDAP connection, which is \fBorg.lsc.jndi.SimpleJndiSrcService\f1. 
.TP
\fBlsc.tasks.\f1TASKNAME\fB.dstService.baseDn\f1
The RDN of the people tree on the destination service. Generally, it should be \fBou=people\f1. 
.TP
\fBlsc.tasks.\f1TASKNAME\fB.dstService.attrId\f1
The attribute used to identify a user in the destination service. This attribute must be used in the identity LDAP filter (see filterId below). For example, if users have a unique value of the LDAP attribute "uid", then you could use the value \fBuid\f1 here. 
.TP
\fBlsc.tasks.\f1TASKNAME\fB.dstService.filterAll\f1
This is the "global LDAP filter" used to retrieve all user DNs from the destination service. 
.TP
\fBlsc.tasks.\f1TASKNAME\fB.dstService.filterId\f1
This is the "identity LDAP filter" used to retrieve one entry from the destination service. You have to use the string \fB{0}\f1 in the filter. This string represents a user identifier found in user entry returned by the global LDAP filter. So, for example, a filter could be \fB(&(objectclass=inetOrgPerson)(mail=*)(uid={0}))\f1
.TP
\fBlsc.tasks.\f1TASKNAME\fB.dstService.attrs\f1
Here, you specify all returned attributes from the destination service one LDAP entry. Generally, these attributes will be used to build the new SASL userPassword. Be careful that the value of this option has to be same as in the source code of this instance of LSC. In fact, the algorithm used is in the JAVA BEAN, in the method named by attributes specified here. So, it strictly not recommended to modify this value (\fBloginName userPassword\f1). 
.SH SYNCHRONIZATION RULE OPTIONS
This section describes synchronization rule options available for this instance of LSC. You should replace \fBTASKNAME\f1 by a task name specified in the global configuration option \fBlsc.tasks\f1. 
.TP
\fBlsc.syncoptions.\f1TASKNAME 
This is the full name of the JAVA class used to provide synchronization option mechanism. For this instance of LSC, you must use the \fBorg.lsc.beans.syncoptions.PropertiesBasedSyncOptions\f1 value. 
.TP
\fBlsc.syncoptions.\f1TASKNAME\fB.default.action\f1
Default action on the destination directory. Here, two values are possible, \fBK\f1 to keep data (means no modification at all), or \fBF\f1 for allowing modification on the data (update, create or delete). For this instance of LSC, you must use \fBK\f1 because source and destination service are in fact the same real server. 
.TP
\fBlsc.syncoptions.\f1TASKNAME\fB.userPassword.action\f1
Choose the action to use onto the userPassword LDAP attribute. As default action, you could put \fBK\f1 to keep password updates, or \fBF\f1 to force modifications. 
.SH AUTHORS
lsc-passwords was written by Sebastien Bahloul <sbahloul@lsc-project.org>, Jonathan Clarke <jclarke@lsc-project.org>, Remy-Christophe Schermesser <rschermesser@lsc-project.org>, Thomas Chemineau <tchemineau@lsc-project.org>. 
.SH SEE ALSO
\fBlsc-agent(1)\f1 \fBlsc.xml(5)\f1

\fBhttp://lsc-project.org/\f1
.SH COMMENTS
This man page was written using \fBxmltoman(1)\f1. 


CHANGES
===== Ldap Synchronization Connector (LSC) Change Log =====

This file contains a summary of changes between releases. Please refer to
individual issues on http://tools.lsc-project.org/ for details.

Please note that this file contains changes for both the lsc-sample and lsc-core
components.


=== LSC release 1.2.1 (15/07/2010) ===

This is a minor upgrade from 1.2.0 that's only purpose is to fix bugs. Users running 1.2.0 or any 1.2-SNAPSHOT version are strongly encouraged to upgrade as soon as possible.

Resolved issues:
- Bug #209: Bad URL parsing for URL with spaces and non ASCII characters
- Bug #216: SLF4J error message: loggers created during the default configuration phase
- Bug #223: srcBean is not consistent when using JavaScript to create a DN
- Bug #224: Bad handling of / character in RDN
- Bug #234: The ldap.search method causes continuation referrals to be thrown
- Bug #249: Tests JAR is not packaged in dist archive - sample can fail in some cases
- Bug #250: Possible NPE in LscAttributes
- Bug #251: logOperation parameter is ignored by CSV logger
- Bug #252: Error deleting entry with a suffix containg a space
- Bug #253: Bad default character handling for cron jobs
- Bug #257: Extra commas returned in ldap.search method causing errors

Bug #215 has also been worked on, allowing the sample to work on Windows in Cygwin.

Please note that this version adds a new dependency on the UnboundID LDAP SDK. The dependency is distributed as a .jar file in the lib/ directory.


=== LSC release 1.2.0 (16/05/2010) ===

If you're upgrading from 1.1.*, see upgrading notes at:
http://lsc-project.org/wiki/documentation/1.2/upgrade-from-1.1

Major changes:
- Generation is no longer necessary (Objects no longer exist, and the new SimpleBean can be used in place of most specific beans)
- LSC-sample no longer exists (a demonstration configuration is in sample/ from the new distribution)
- Ant and Maven are no longer required. The only dependency for LSC is JRE (other dependencies are bundled)
- db2ldap configuration has been simplified: The configuration file database.properties is obsolete. Database connection properties are now read from the main lsc.properties file.
- The logging system and configuration has changed: logging configuration is now done via the logback.xml file.

Resolved issues:
- Feature #22: Don't require generation for attributes and objectclasses from the schema in the directory
- Feature #33: Support attribute names containing a "-" character
- Feature #59: Implement post hooks after sync and clean
- Feature #103: Create a distribution tarball from lsc-core including quickstart possibilities
- Feature #112: Upgrade to OpenDS 1.2 stable version (for the sample)
- Feature #136: Plug in any source connector easily
- Feature #139: Migrate from log4j to sfl4j & LogBack
- Feature #153: AntInstaller deprecated
- Feature #154: Migrate the stack traces from LOGGER.error to LOGGER.debug
- Feature #187: ldap.exists caller method
- Feature #198: Display warning message when configuration file not found
- Feature #214: All deprecated methods should log warnings
- Bug #32: [ldap2ldap] Source objectClass is created in destination directory
- Bug #99: getContextDn() is not found by rhino
- Bug #107: Null values not supported in CSV files
- Bug #140: JScriptEvaluator method evalToStringList() may return a list of one empty string
- Bug #157: Standardise output parameters for logging configuration
- Bug #168: lsc fails with unrelevant message due to bad config file
- Bug #170: Correct default logging settings
- Bug #174: Update sample for 1.2
- Bug #176: add function getAccountExpires to AD.class
- Bug #179: Quickstart tutorial fails
- Bug #184: create_value inconsistency between 1.1.2 and 1.2
- Bug #190: utf-8 and import csv
- Bug #191: ScriptableJndiServices exists method ignores filter
- Bug #192: Message about cygwin path under GNU/Linux
- Bug #196: Ldif layout bug for modrdn LDAP operation
- Bug #197: Problem with DN that has a # sign
- Bug #202: LSC silently fails if the key for SymmetricEncryption can't be read
- Bug #203: Error in documentation for SymmetricEncryption: property keypath doesn't exist, it's keyfile
- Bug #204: EvaluatorException for value containing ";"
- Bug #115: After OpenDS 2.0 upgrade, directory fails to shutdown (development)


=== LSC release 1.1.2 (07/12/2009) ===

- Fixed Bug #141: Create and default values not added to destination for attributes not in source object
- Fixed Bug #145: db2ldap synchronizations with multiple pivot attributes don't handle all entries
- Fixed Bug #148: Quickstart tutorial was broken (http://lsc-project.org/wiki/documentation/quickstart)
- Fixed Bug #151: Null Pointer Exception in logActionError if errors occured very early
- Fixed Bug #152: Add and modify operations do not handle empty fields the same way


=== LSC release 1.1.1 (01/11/2009) ===

- Fixed Bug #118: LDIF logger throws exception on zero length value
- Fixed Bug #119: SQLMap url in example sql-map-config.xml is wrong
- Fixed Bug #128: Bad value checking for default values in BeanComparator
- Added Feature #97: Explicit error message when Rhino fails


=== LSC release 1.1.0 (27/07/2009) ===

- Initial release of LSC 1.1.
- Changes from the 1.0 branch are too numerous to list.


src/main/java/org/lsc/configuration/Lsc.java
//
// This file was generated by the JavaTM Architecture for XML Binding(JAXB) Reference Implementation, v2.3.0 
// See <a href="https://javaee.github.io/jaxb-v2/">https://javaee.github.io/jaxb-v2/</a> 
// Any modifications to this file will be lost upon recompilation of the source schema. 
// Generated on: 2021.08.24 at 10:00:27 AM CEST 
//


package org.lsc.configuration;

import javax.xml.bind.annotation.XmlAccessType;
import javax.xml.bind.annotation.XmlAccessorType;
import javax.xml.bind.annotation.XmlAttribute;
import javax.xml.bind.annotation.XmlElement;
import javax.xml.bind.annotation.XmlRootElement;
import javax.xml.bind.annotation.XmlType;


/**
 * <p>Java class for anonymous complex type.
 * 
 * <p>The following schema fragment specifies the expected content contained within this class.
 * 
 * <pre>
 * &lt;complexType&gt;
 *   &lt;complexContent&gt;
 *     &lt;restriction base="{http://www.w3.org/2001/XMLSchema}anyType"&gt;
 *       &lt;all&gt;
 *         &lt;element name="connections" type="{http://lsc-project.org/XSD/lsc-core-2.2.xsd}connectionsType"/&gt;
 *         &lt;element name="audits" type="{http://lsc-project.org/XSD/lsc-core-2.2.xsd}auditsType" minOccurs="0"/&gt;
 *         &lt;element name="tasks" type="{http://lsc-project.org/XSD/lsc-core-2.2.xsd}tasksType"/&gt;
 *         &lt;element name="security" type="{http://lsc-project.org/XSD/lsc-core-2.2.xsd}securityType" minOccurs="0"/&gt;
 *       &lt;/all&gt;
 *       &lt;attribute name="id" type="{http://www.w3.org/2001/XMLSchema}string" /&gt;
 *       &lt;attribute name="revision" type="{http://www.w3.org/2001/XMLSchema}int" /&gt;
 *     &lt;/restriction&gt;
 *   &lt;/complexContent&gt;
 * &lt;/complexType&gt;
 * </pre>
 * 
 * 
 */
@XmlAccessorType(XmlAccessType.FIELD)
@XmlType(name = "", propOrder = {

})
@XmlRootElement(name = "lsc")
public class Lsc {

    @XmlElement(required = true)
    protected ConnectionsType connections;
    protected AuditsType audits;
    @XmlElement(required = true)
    protected TasksType tasks;
    protected SecurityType security;
    @XmlAttribute(name = "id")
    protected String id;
    @XmlAttribute(name = "revision")
    protected Integer revision;

    /**
     * Gets the value of the connections property.
     * 
     * @return
     *     possible object is
     *     {@link ConnectionsType }
     *     
     */
    public ConnectionsType getConnections() {
        return connections;
    }

    /**
     * Sets the value of the connections property.
     * 
     * @param value
     *     allowed object is
     *     {@link ConnectionsType }
     *     
     */
    public void setConnections(ConnectionsType value) {
        this.connections = value;
    }

    /**
     * Gets the value of the audits property.
     * 
     * @return
     *     possible object is
     *     {@link AuditsType }
     *     
     */
    public AuditsType getAudits() {
        return audits;
    }

    /**
     * Sets the value of the audits property.
     * 
     * @param value
     *     allowed object is
     *     {@link AuditsType }
     *     
     */
    public void setAudits(AuditsType value) {
        this.audits = value;
    }

    /**
     * Gets the value of the tasks property.
     * 
     * @return
     *     possible object is
     *     {@link TasksType }
     *     
     */
    public TasksType getTasks() {
        return tasks;
    }

    /**
     * Sets the value of the tasks property.
     * 
     * @param value
     *     allowed object is
     *     {@link TasksType }
     *     
     */
    public void setTasks(TasksType value) {
        this.tasks = value;
    }

    /**
     * Gets the value of the security property.
     * 
     * @return
     *     possible object is
     *     {@link SecurityType }
     *     
     */
    public SecurityType getSecurity() {
        return security;
    }

    /**
     * Sets the value of the security property.
     * 
     * @param value
     *     allowed object is
     *     {@link SecurityType }
     *     
     */
    public void setSecurity(SecurityType value) {
        this.security = value;
    }

    /**
     * Gets the value of the id property.
     * 
     * @return
     *     possible object is
     *     {@link String }
     *     
     */
    public String getId() {
        return id;
    }

    /**
     * Sets the value of the id property.
     * 
     * @param value
     *     allowed object is
     *     {@link String }
     *     
     */
    public void setId(String value) {
        this.id = value;
    }

    /**
     * Gets the value of the revision property.
     * 
     * @return
     *     possible object is
     *     {@link Integer }
     *     
     */
    public Integer getRevision() {
        return revision;
    }

    /**
     * Sets the value of the revision property.
     * 
     * @param value
     *     allowed object is
     *     {@link Integer }
     *     
     */
    public void setRevision(Integer value) {
        this.revision = value;
    }

}
